### DISCUSSION TITLE
New API Integration

### DISCUSSION THREAD

**Discussion by mfleader** [Hyperfoil/Horreum#629] [General]
I am interested in creating a Golang REST client to interact with Horreum's HTTP api. 

I've never used Quarkus before, so I am a little unfamiliar with how it is being used in your codebase to generate api endpoints. I am trying to use the horreum-client directory and Horreum's generated OpenAPI schema to determine the required data classes (models) for a client, but I am a little confused with many of the imports from RestEasy and JaxRS, and specifically JsonNode. 

If any of this discussion is too far afield from Horreum, like JsonNode and how or if I need to implement all of its functionality in Golang, please let me know!

**Comment by jesperpedersen** [comment]
Did you look at https://github.com/deepmap/oapi-codegen ?

**Reply by mfleader** [reply]
Yes, the code generation failed.

```shell
❯ oapi-codegen -package horreum ~/horreum.yaml > horreum.gen.go
error generating code: error creating operation definitions: 
error describing global parameters for GET//api/action/list: 
error generating type for param (direction): error merging schemas: 
error merging schemas for AllOf: merging two sets of defaults is undefined
```

**Comment by mfleader** [comment]
I'm curious about JsonNode because I think the OpenAPI schema is interpreting as a sort of "any" type, as in [Run data](https://github.com/Hyperfoil/Horreum/blob/dc14b7377ca6dc6e0987dac6d4345c48fa3ceddd/horreum-api/src/main/java/io/hyperfoil/tools/horreum/api/data/Run.java#L23). 

My secondary reason for interest in it is because I think the [OpenAPI Golang generator](https://openapi-generator.tech/#try)  does not know how to create a 'list of any' type.

**Comment by johnaohara** [comment]
Hi @mfleader, there are a number of Jackson Types that are leaking from the API atm, there are `ObjectNode`, `JsonNode`, and `ArrayNode`.  We are currently working to remove them a replace them with concrete types. 

The openapi spec file for Horreum is automatically generated at build time and is found at `horreum-api/target/generated/openapi.*`

Can i ask where you found the `horreum.yaml` file?

**Comment by jesperpedersen** [comment]
@mfleader It would be great if you could help us lock the API down so it will easier to integrate with non-Java based programming languages

Could you try going through our https://github.com/Hyperfoil/Horreum/blob/master/README.md to setup it up locally ?

**Comment by mfleader** [comment]
During validation the openapi generator says the spec has errors

```shell
❯ npx @openapitools/openapi-generator-cli generate -i ~/horreum.yaml -g python --git-repo-id 680226051 --git-user-id 65633602
Exception in thread "main" org.openapitools.codegen.SpecValidationException: There were issues with the specification. The option can be disabled via validateSpec (Maven/Gradle) or --skip-validate-spec (CLI).
 | Error count: 20, Warning count: 0
Errors: 
	-attribute components.schemas.ExperimentResult.items is missing
	-attribute components.schemas.ViewComponent.items is missing
	-attribute components.schemas.ValidationError.items is missing
	-attribute components.schemas.ExperimentComparison.items is missing
	-attribute components.schemas.Run.items is missing
	-attribute components.schemas.ChangeDetection.items is missing
	-attribute components.schemas.RunExtended.items is missing
	-attribute components.schemas.TableReportData.items is missing
	-attribute components.schemas.LabelValue.items is missing
	-attribute components.schemas.LabelPreview.items is missing
	-attribute components.schemas.Dataset.items is missing
	-attribute components.schemas.DatasetSummary.items is missing
	-attribute components.schemas.RunSummary.items is missing
	-attribute components.schemas.Variable.items is missing
	-attribute paths.'/api/run/data'(post).requestBody.content.'application/json'.schema.items is missing
	-attribute components.schemas.ReportComponent.items is missing
	-attribute components.schemas.Action.items is missing
	-attribute components.schemas.Schema.items is missing
	-attribute components.schemas.MissingDataRule.items is missing
	-attribute components.schemas.TableReportConfig.items is missing

	at org.openapitools.codegen.config.CodegenConfigurator.toContext(CodegenConfigurator.java:668)
	at org.openapitools.codegen.config.CodegenConfigurator.toClientOptInput(CodegenConfigurator.java:695)
	at org.openapitools.codegen.cmd.Generate.execute(Generate.java:503)
	at org.openapitools.codegen.cmd.OpenApiGeneratorCommand.run(OpenApiGeneratorCommand.java:32)
	at org.openapitools.codegen.OpenAPIGenerator.main(OpenAPIGenerator.java:66)
```

The log errors when I use the `--skip-validate-spec` option are long.

**Reply by mfleader** [reply]
I feel like this is a bug with the OpenAPI Generator's validation. The python package [openapi spec validator](https://github.com/python-openapi/openapi-spec-validator) and the web application API Tools [Swagger parser](https://apitools.dev/swagger-parser/online) both agree that your web hosted [horreum spec](https://horreum.hyperfoil.io/docs/reference/openapi.yaml) is valid.

**Comment by mfleader** [comment]
In the [horreum spec](https://horreum.hyperfoil.io/docs/reference/openapi.yaml), why is the web application api mixed in with the data storage and analysis api?

**Reply by johnaohara** [reply]
Hi @mfleader, myseld and @stalep were discussing this last week, separating the different target clients into different api schemas, so cli/analytical tools do not have to consume the web UI api

**Reply by mfleader** [reply]
I did not realize that the cli and analytical tools would have to consume the web api. 

Does that mean it is not possible to interact with horreum without going through the web ui?

**Reply by jesperpedersen** [reply]
@mfleader "do **not** have to" :) We are looking at making the REST API stable

**Reply by mfleader** [reply]
Does that mean the API to Horreum's data services is exposed for use?

**Comment by stalep** [comment]
@mfleader we've now split the -api into two modules; horreum-ui now contains most of the data/analytics services and models and horreum-api is more "client" focused. We have verified that both python and go clients can be created with the generated yaml file based on horreum-api.
We're going for a new release fairly soonish and we hope this will solve the issues we've had with the client generation.

**Reply by mfleader** [reply]
That's awesome! Solid work y'all

**Comment by stalep** [comment]
I'm closing this as the API is compatible with the spec and can be found here https://horreum.hyperfoil.io/openapi/
